lecture-notes/2-vcs.md

```diff
@@ -1764,7 +1764,205 @@ decided to give it its own subcommand, and `git restore` was born.
+use your newfound `git checkout` skills to do so, you might see something like
+this:
+
+```console?prompt=$
+git checkout 2d6603026105
+error: Your local changes to the following files would be overwritten by checkout:
+	main.c
+Please commit your changes or stash them before you switch branches.
+Aborting
+$ 
+```
+
+The cause of this is your half-finished refactor, which you haven't committed
+yet:
+
+```diff
+$ git diff
+diff --git a/main.c b/main.c
+index 56ef02d..6f0fa39 100644
+--- a/main.c
++++ b/main.c
+@@ -1,6 +1,7 @@
+ #include <errno.h>
+ #include <stdio.h>
+ #include <stdlib.h>
++#include <stdbool.h>
+
+ int calc(int left, char op, int right) {
+   switch (op) {
+@@ -17,6 +18,23 @@ int calc(int left, char op, int right) {
+   exit(EXIT_FAILURE);
+ }
+
++// Converts the string str into an integer using strtol.
++// Returns 1 and stores the result in resultp on success.
++// Returns 0 and sets errno otherwise.
++bool to_int(const char *str, int *resultp) {
++  char *endptr;
++  errno = 0;
++  int result = strtol(str, &endptr, 10);
++  if (str == endptr) {
++    errno = EINVAL;
++  }
++  if (errno != 0) {
++    return false;
++  }
++  *resultp = result;
++  return true;
++}
++
+ int main(int argc, char **argv) {
+   if (argc != 4) {
+     fprintf(stderr,
+@@ -29,11 +47,8 @@ int main(int argc, char **argv) {
+   const char *right_str = argv[3];
+   errno = 0;
+   char *endptr;
+-  int left = strtol(left_str, &endptr, 10);
+-  if (left_str == endptr) {
+-    errno = EINVAL;
+-  }
+-  if (errno != 0) {
++  int left;
++  if (!to_int(left_str, left)) {
+     perror(argv[0]);
+     return EXIT_FAILURE;
+   }
+```
+
+Git sees that you've modified `main.c`, so it refuses to check out any other
+version of that file for fear of destroying your modifications! Although this is
+a nice safety feature, it can be annoying: you're not ready to commit your
+refactor---it crashes and introduces a compiler warning---but you have more
+important things to do than debug it. So what can you do?
+
+One option is to commit the unfinished change and finish it in a later commit.
+Although this violates our rule that each commit should be complete and
+self-contained, it can be a viable strategy when the refactoring work is on its
+own branch. (More on that next lecture.) Another is to destroy your work so far
+(e.g. by `git restore`ing `main.c`) and redo it later, which is clearly
+undesirable.
+
+A third option is to use a subcommand called `git stash` to save your work for
+later. `git stash` is a bit like `git commit`, in that it saves changes from the
+working tree to a more permanent location. Unlike `git commit`, however, those
+changes don't become part of the Git history. Instead, they go into a special
+area called the *stash*, which unlike the history is neither linear nor shared
+between different copies of the repository. `git stash` is shorthand for `git stash push`, which moves all uncommitted changes to a new stash entry:
```

Comment:
worth calling out that it is a stack of commits, and therefore snapshots, but its primary mechanism is even moreso to look like a stack of patches (push+pop/apply)